prime-router/src/main/kotlin/Submission.kt

```diff
@@ -186,20 +193,131 @@ class DetailedSubmissionHistory(
             }
         }
     }
+
+    /**
+     * Consolidate the [logs] filtered by an optional [filterBy] action level, so to list similar messages once
+     * with a list of items they relate to.
+     * @return the consolidated list of logs
+     */
+    internal fun consolidateLogs(filterBy: ActionLogLevel? = null):
+        List<ConsolidatedActionLog> {
+        val consolidatedList = mutableListOf<ConsolidatedActionLog>()
+        if (logs != null) {
+            // First filter the logs and sort by the message.  This first sorting can take care of sorting old messages
+            // that contain index numbers like "Report 3: xxxx"
+            val filteredList = when (filterBy) {
+                null -> logs
+                else -> logs.filter { it.type == filterBy }
+            }.sortedBy { it.detail.message }
+            // Now order the list so that logs contain first non-item messages, then item messages, and item messages
+            // are sorted by index.
+            val orderedList = (
+                filteredList.filter { it.scope != ActionLogScope.item }.sortedBy { it.scope } +
+                    filteredList.filter { it.scope == ActionLogScope.item }.sortedBy { it.index }
+                ).toMutableList()
+            // Now consolidate the list
+            while (orderedList.isNotEmpty()) {
+                // Grab the first log.
+                val consolidatedLog = ConsolidatedActionLog(orderedList.first())
+                orderedList.removeFirst()
+                // Now find similar logs and consolidate it.  Note by using the iterator we can delete on the fly.
+                with(orderedList.iterator()) {
+                    forEach { log ->
+                        if (consolidatedLog.canBeConsolidatedWith(log)) {
+                            consolidatedLog.add(log)
+                            remove()
+                        }
+                    }
+                }
+                consolidatedList.add(consolidatedLog)
+            }
+        }
+        return consolidatedList
+    }
+}
+
+/**
+ * Consolidated action log class to be output to the API JSON response.
+ * @param log the base log message to be consolidated
+ */
+@JsonInclude(Include.NON_NULL)
+class ConsolidatedActionLog(log: DetailActionLog) {
+    /**
+     * The scope of the log.
+     */
+    var scope: ActionLogScope
+
+    /**
+     * The list of indices for item logs. An index can be null if there was no index provided with the log.
+     */
+    var indices: MutableList<Int?>? = null
+
+    /**
+     * The list of tracking IDs for item logs. A tracking ID can be null if there was no ID provided with the log.
+     */
+    var trackingIds: MutableList<String?>? = null
+
+    /**
+     * The log level.
+     */
+    var type: ActionLogLevel
+
+    /**
+     * The field mapping for item logs.
+     */
+    var field: String? = null
+
+    /**
+     * The log message.
+     */
+    var message: String
```

Comment:
Can these be made into non-null val's?

Reply:
Very correct.  It can be val.  Not sure why Intellij did not complain.